### PR TITLE
In memory audio blob loading and splitting

### PR DESCRIPTION
**Pull request by rlancemartin**
This PR creates a `BaseBlobSplitter` and a `BaseBlobTransformer`, which can be used to split and transform blobs. It creates an instance of the `BaseBlobTransformer` called `YoutubeToAudioTransformer`, which can be used to download a YouTube URL to audio. In addition, it creates an instance of `BaseBlobSplitter` called `AudioSplitter`, which can be used to split an audio file into smaller blobs. Both are useful for an end-to-end workflow that combines (1) YouTube link to Audio, (2) Audio to Audio splits, and (3) then use `OpenAIWhisperParser` [here](https://github.com/hwchase17/langchain/pull/5580) to create `Documents` from the splits.

### CODE REVIEW

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from langchain.document_loaders.base import BaseBlobTransformer
+from langchain.document_loaders.blob_loaders import Blob
+
+
+class YoutubeToAudioTransformer(BaseBlobTransformer):
```
Comment (eyurtsev) [inline langchain/document_loaders/blob_transformers/youtube_to_audio.py:5]: VideoToAudioTransformer

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+from langchain.document_loaders.base import BaseBlobTransformer
```
Comment (eyurtsev) [inline langchain/document_loaders/blob_transformers/youtube_to_audio.py:1]: We should figure out how to organize the files. 

Here, we're dealing with classes that are called blob transformers which transform binary data to binary data of other format.

These are very similar to parsers. Parsers transform blobs to documents. Documents can be viewed as a blob that's been loaded into memory and looks like textual data.

I don't have a strong opinion at the moment, but I think we should think for a bit about the name spacing to make the code easy to discover (and relatively fast to import!)

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+class YoutubeToAudioTransformer(BaseBlobTransformer):
+
+    """Dump YouTube url as audio file."""
+
+    # TODO: Output should be the blob path?
+    def lazy_transform(self, input_url: Blob) -> Blob:
+        """Lazily transform the blob."""
+
+        import yt_dlp
```
Comment (eyurtsev) [inline langchain/document_loaders/blob_transformers/youtube_to_audio.py:13]: We can move this code into a blob generator, which accepts a list of urls and yields video blobs.

**Comment by eyurtsev** [comment]
@rlancemartin looking great, left suggestions to rename a few things. 

I want to propose something radical -- let's take out all file I/O for the first iteration, so there's no file artifacts generated during the entire process and instead everything is streaming in memory.

Did you add file IO to avoid re-downloading / re-processing content or for another reason?

File: langchain/document_loaders/base.py [hunk 4]
```diff
@@ -85,3 +85,85 @@ def parse(self, blob: Blob) -> List[Document]:
...
+
+    A blob transformer provides a way to transform raw data stored in a blob into a
+    different data format.
+
+    The transformer can be composed with blob loaders, making it easy to re-use
+    a transformer independent of how the blob was originally loaded.
+    """
+
+    @abstractmethod
+    def lazy_transform(self, blob: Blob) -> Blob:
```
[the remark this answers is not in the document]
Comment (rlancemartin) [reply langchain/document_loaders/base.py:101]: got it! thanks for the clarification on that.

File: langchain/document_loaders/base.py [hunk 5]
```diff
@@ -85,3 +85,85 @@ def parse(self, blob: Blob) -> List[Document]:
...
+            Blob
+        """
+        return self.lazy_transform(blob)
+
+
+class BaseBlobSplitter(ABC):
+    """Abstract interface for blob splitters.
+
+    A blob splitter provides a way to transform raw data stored in a blob into a
+    different data format.
```
[the remark this answers is not in the document]
Comment (rlancemartin) [reply langchain/document_loaders/base.py:135]: yes, this was a copy-paste error :)

File: langchain/document_loaders/blob_splitters/split_audio.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    # TODO: Output should be a list of the blob paths?
+    def lazy_split(self, blob: Blob) -> Iterator[Blob]:
+        """Lazily split the blob."""
+
+        from pydub import AudioSegment
+
+        # TODO: Determine best way to pass this
+        output_file_path = "path_to_files"
+
+        video = AudioSegment.from_file(blob)
```
[the remark this answers is not in the document]
Comment (rlancemartin) [reply langchain/document_loaders/blob_splitters/split_audio.py:20]: yes, good call out: `blob.path` should work from having a quick look at `blob_loaders/schema.py`

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py (hunk 1 above)
Comment (rlancemartin) [reply langchain/document_loaders/blob_transformers/youtube_to_audio.py:5]: one subtle point here: `yt_dlp` takes a YouTube url specifically. 

so the functionality more specific than `VideoToAudioTransformer`.

the flow here is something like:
1/ we can add a `WebUrlBlobLoader` (as you note that we should add [here](https://www.notion.so/LangChain-Documents-786e0b2762694ca786889556ea4586ba#1de4c8a6cedf4aaab8d501e32afe4b81).
2/ we then extract the `url` from the blob and pass it here

File: langchain/document_loaders/blob_splitters/split_audio.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        # Define the duration of each chunk in minutes
+        chunk_duration = 20
+
+        # Calculate the chunk duration in milliseconds
+        chunk_duration_ms = chunk_duration * 60 * 1000
+
+        # Split the audio into chunk_duration-minute chunks
+        for i in range(0, len(video), chunk_duration_ms):
+            chunk = video[i : i + chunk_duration_ms]
+            fpath = output_file_path + f"audio_chunk_{i // chunk_duration_ms}.mp3"
```
[the remark this answers is not in the document]
Comment (rlancemartin) [reply langchain/document_loaders/blob_splitters/split_audio.py:31]: +1, im modifying the code now.

one example flow we want to support:
1/ fetch url (TODO): `GenericLoader.from_url`
2/ transcribe audio: `YoutubeToAudioTransformer`
3/ split audio: `AudioSplitter`
4/ transcribe audio to docs ([PR](https://github.com/hwchase17/langchain/blob/ac00b73a8d64b373d4d19efa543be66b79b32aa9/docs/modules/indexes/document_loaders/examples/audio.ipynb)): `OpenAIWhisperParser `

the in memory flow would simply be:
```
GenericLoader.from_url(youtube_url).YoutubeToAudioTransformer().AudioSplitter(parser=OpenAIWhisperParser())
```

im not sure the `OpenAIWhisperParser` will work w/ `AudioSplitter` in its current form.

but i will made code changes that get us close to this, and we can discuss.

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py (hunk 2 above)
Comment (rlancemartin) [reply langchain/document_loaders/blob_transformers/youtube_to_audio.py:1]: actually, i like your suggestion about the in memory flow.

in that case, we won't have to worry about writing these intermediate files.

i'll share that code shorty.

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py (hunk 3 above)
Comment (rlancemartin) [reply langchain/document_loaders/blob_transformers/youtube_to_audio.py:13]: minor nit on this: `yt_dlp` downloads from url to audio directly w/o the video file intermediate.

Comment (eyurtsev) [reply langchain/document_loaders/blob_transformers/youtube_to_audio.py:13]: Ah I see!

File: langchain/document_loaders/blob_transformers/youtube_to_audio.py (hunk 1 above)
Comment (eyurtsev) [reply langchain/document_loaders/blob_transformers/youtube_to_audio.py:5]: Makes sense! 

We could either use a specific Blob Loader or a more generic WebURLBlobLoader that takes in a download handler. 

The former could be simpler as a first pass so we don't have to think about how to implement a generic download handler etc

```python
class YoutubeAudioLoader(BlobLoader)?
  def __init__(urls):
    ...

  def yield_blobs() -> Iterable[Blob]: # <-- Yields Blobs that have data attribute populated with the audio
   pass
```

File: langchain/document_loaders/parsers/audio.py [hunk 8]
```diff
@@ -12,9 +12,18 @@ class OpenAIWhisperParser(BaseBlobParser):
...
+        import io 
         import openai
 
+        # Need if we pass an in-memory blob from AudioSplitter().lazy_split(video)
+        file_obj = io.BytesIO(blob.data)
+        file_obj.name = "input_split.mp3"
+
+        # This fails when split is not read from disk (e.g., w/ GenericLoader)
+        # OAI API is missing "f.name"
         with blob.as_bytes_io() as f:
```
Comment (eyurtsev) [inline langchain/document_loaders/parsers/audio.py:24]: Let's create a buffered reader here
```
from io import BytesIO, BufferedReader
b = io.BytesIO(b'some data')
b.name = "hello.mp3"
b.seek(0)
reader = BufferedReader(b)
```

We can then remove the context manager that uses `blob.as_bytes_io()`, so we avoid reading the blob data twice.

---

We can follow up later and add this capability to the blob object to allow it to yield a named buffered reader.

File: langchain/document_loaders/parsers/audio.py [hunk 9]
```diff
@@ -12,9 +12,18 @@ class OpenAIWhisperParser(BaseBlobParser):
     def lazy_parse(self, blob: Blob) -> Iterator[Document]:
         """Lazily parse the blob."""
 
+        import io 
         import openai
 
+        # Need if we pass an in-memory blob from AudioSplitter().lazy_split(video)
+        file_obj = io.BytesIO(blob.data)
+        file_obj.name = "input_split.mp3"
```
Comment (eyurtsev) [inline langchain/document_loaders/parsers/audio.py:20]: Is this name important for anything in terms of how the openai API uses it? Is it OK that multiple audio files share the same filename?

File: langchain/document_loaders/blob_splitters/split_audio.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+from typing import Iterator
```
Comment (eyurtsev) [inline langchain/document_loaders/blob_splitters/split_audio.py:1]: Looks great!

**Comment by eyurtsev** [comment]
@rlancemartin left a few comments for minor changes, we're looking pretty good -- let me know if makes sense -- should be able to implement and then re-lint and we can merge

**Comment by rlancemartin** [comment]
We merged a refactor [here](https://github.com/hwchase17/langchain/pull/5772/files): 1/ audio loading in memory will require something other than `yt_dlp`. 2/ exposing the splitter is probably not required b/c there are no obvious use-cases for it; it's more of an internal transformation for the parser due to the OpenAI Whisper API side limit.

**Pull request closed without merging**